crates/provider/src/chain/stream.rs

```diff
@@ -0,0 +1,159 @@
+            let new_number = match self.block_numbers_stream.next().await {
+                Some(Ok(block_number)) => block_number.to::<u64>(),
+                Some(Err(err)) => {
+                    // This is fine.
+                    debug!(%err, "polling stream lagged");
+                    continue;
+                }
+                None => {
+                    debug!("polling stream ended");
+                    return Err(ChainStreamError::PollingStreamEnded);
+                }
+            };
+            self.chain_tip = new_number;
+        }
+        Ok(())
+    }
+
+    /// Tries to fetch the range of blocks that can surely be yielded via single batch request.
+    async fn fetch_blocks(&self) -> Result<Vec<Block>, ChainStreamError<T>> {
+        // We don't want to request too many batches by accident.
+        // TODO: Should be a property on transport or client.
+        const MAX_BATCH_SIZE: usize = 10;
+
+        let range_end = self.chain_tip.min(self.next_yield + MAX_BATCH_SIZE as u64 - 1);
+        let range = self.next_yield..=range_end;
+        if range.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Perform a batch request.
+        let client = Self::client(&self.client)?;
+        let mut batch_request = BatchRequest::new(client.as_ref());
+        let futures: FuturesOrdered<Waiter<Option<Block>>> = range
+            .map(|number| {
+                batch_request
+                    .add_call::<_, Option<Block>>(
+                        "eth_getBlockByNumber",
+                        &(U64::from(number), false),
+                    )
+                    .expect("Cannot serialize call params")
+            })
+            .collect();
+        batch_request.send().await.map_err(ChainStreamError::Rpc)?;
+
+        // We expect all the futures to complete simultaneously, given that it's a batch request,
+        // so we will collect them and let the caller yield the results.
+        let mut blocks = Vec::with_capacity(futures.len());
+        for maybe_block in futures.collect::<Vec<_>>().await {
+            match maybe_block {
+                Ok(Some(block)) => {
+                    debug!(number = self.next_yield, "yielding block");
+                    blocks.push(block)
+                }
+                Ok(None) => {
+                    // We expect that the remaining blocks will be missing too.
+                    break;
+                }
+                Err(err) => {
+                    error!(%err, "failed to fetch blocks");
+                    return Err(ChainStreamError::Rpc(err));
+                }
+            }
+        }
+        Ok(blocks)
+    }
+
+    /// Performs a single step of the polling loop.
+    /// Either returns a list of blocks to yield, or an irrecoverable error.
+    async fn new_blocks(&mut self) -> Result<Vec<Block>, ChainStreamError<T>> {
+        self.update_chain_tip().await?;
+        self.fetch_blocks().await
```

Comment:
Yes, that's deliberate. I think a "fully parallel" solution would be significantly more complex, and not sure if that's justified here -- in most realistic cases the blocks will be consumed one by one right after they're fetched.
This solution is more efficient than the previous one, though, since we don't have to wait for the underlying poller if we know that we're lagging behind.